### PR TITLE
Faster and more accurate sinpi, cospi, sincospi for Float32, Float64

### PR DESCRIPTION
**Pull request by oscardssmith**
Instead of using `sin_kernel` and `cos_kernel`, use better kernels that need less precision.

I benchmark this as 14% faster for Float64, 7% faster for Float32.

Accuracy for Float64 is .64 ULPs compared to .75 for master.

Accuracy for Float32 is .5 ULP

### CODE REVIEW

File: base/special/trig.jl [hunk 1]
```diff
@@ -740,6 +740,38 @@ mulpi_ext(x::Float32) = DoubleFloat32(pi*Float64(x))
...
+    r  = evalpoly(x², (2.5501640398773415, -0.5992645293202981, 0.08214588658006512,
+					   -0.007370429884921779, 0.0004662827319453555, -2.1717412523382308e-5))
+	
+    return muladd(3.141592653589793, x, x*muladd(-5.16771278004997, 
+				  x², muladd(x⁴, r,  1.2245907532225998e-16)))
+end
+
+@inline function cospi_kernel(x::Float64)
+    x² = x*x
+	r = x²*evalpoly(x², (4.058712126416765, -1.3352627688537357, 0.23533063027900392,
```
Comment (KristofferC) [inline base/special/trig.jl:755]: Indentation looks strange in this function.

Comment (dkarrasch) [reply base/special/trig.jl:755]: Indentation looks strange throughout, actually, e.g., around (new) line 800 etc.

Comment (oscardssmith) [reply base/special/trig.jl]: oops, tabs got into here by accident.

File: base/special/trig.jl [hunk 2]
```diff
@@ -740,6 +740,37 @@ mulpi_ext(x::Float32) = DoubleFloat32(pi*Float64(x))
 mulpi_ext(x::Rational) = mulpi_ext(float(x))
 mulpi_ext(x::Real) = pi*x # Fallback
 
+@inline function sinpi_kernel(x::Float64)
```
Comment (staticfloat) [inline base/special/trig.jl:743]: Would be kind of nice to have a brief description of the method by which this works, similar to that of `acos()` on line 671.  Just a brief mention of something like "uses a taylor series expansion around 0" or something like that.

**Comment by staticfloat** [comment]
Thanks Oscar.  For my own curiosity, how did you generate these polynomial coefficients?  If it's not too much code, it would be cool to record that for future usage with other special functions, so that you're not the only person who knows how to do this.

Test failures look like `cospi_kernel(::BigFloat)` is trying to be called.

**Comment by oscardssmith** [comment]
The coefficient generation is slightly complex, but the TLDR is to use Remez to generate a minimax polynomial over the range of inputs, specialized to the even/oddness of the polynomial. I really need to put the general principals all into a blogpost at some point, but the TLDR for this is

```
using Remez
sinpi_coeffs = [(big(pi)*x)^n/factorial(big(n)) for n in 1:2:40] #nonzero taylor coefficients in bigfloat
sinpi_kinda(x) = evalpoly(x, sinpi_coeffs)
bounds = (0, 0.25^2) #squared because we are getting coefficients for sqrt(sinpi(x))
n = 7 # degree 7 polynomial
exact_coeficients = ratfn_minimax(sinpi_kinda, bounds, n, 0)[1]
#now do lots of messing around with rounding to get things exact.
```

**Comment by oscardssmith** [comment]
This has been surprisingly tricky to get right. At the moment, the only failing tests is the behavior as to whether `0.0` or `-0.0` is returned. To me, neither is clearly right. Is there a good reason why the specific signs need to be tested? If not, I will probably loosen the tests slightly to check for `==0` rather than `===0.0`

**Comment by oscardssmith** [comment]
adding triage label to discus signs of zeros.

**Comment by oscardssmith** [comment]
removing triage tag after research suggests Base was correct.

**Comment by ViralBShah** [comment]
Is it worth trying to get this done?

**Comment by oscardssmith** [comment]
yeah. it's still on my radar. I just haven't had the time recently.

**Comment by simonbyrne** [comment]
re: signed zeros, the IEEE754-2019 spec states
> sinPi(+n) is +0 and sinPi(−n) is −0 for positive integers n
> cosPi(n + 1⁄2) is +0 for any integer n when n + 1⁄2 is representable

**Comment by oscardssmith** [comment]
I think this is finally ready to go. Anyone want to review?

**Review by gbaraldi — APPROVED** [review]
LGTM

**Pull request merged**